Assets/MixedRealityToolkit-SDK/Features/Audio/Influencers/AudioOccluder.cs

```diff
@@ -0,0 +1,145 @@
+        [Tooltip("Frequency above which sound will not be heard after applying occlusion.")]
+        [Range(10.0f, 22000.0f)]
+        [SerializeField]
+        private float cutoffFrequency = 5000.0f;
+        public float CutoffFrequency
+        {
+            get { return cutoffFrequency; }
+            set
+            {
+                // set cutoffFrequency and enforce the specified range
+                if (value < 10.0f)
+                {
+                    cutoffFrequency = 10.0f;
+                }
+                else if (value > 22000.0f)
+                {
+                    cutoffFrequency = 22000.0f;
+                }
+                else
+                {
+                    cutoffFrequency = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Percentage of the audio source volume that will be heard after applying occlusion.
+        /// </summary>
+        /// <remarks>
+        /// VolumePassThrough is cumulative. It is applied to the current volume of the object at the time
+        /// the effect is applied.
+        /// The VolumePassThrough range is from 0.0 - 1.0 (0-100%), inclusive.
+        /// The default value is 1.0.
+        /// </remarks>
+        [Tooltip("Percentage of the audio source volume that will be heard after applying occlusion.")]
+        [Range(0.0f, 1.0f)]
+        [SerializeField]
+        private float volumePassThrough = 1.0f;
+        public float VolumePassThrough
+        {
+            get { return volumePassThrough; }
+            set
+            {
+                // set cutoffFrequency and enforce the specified range
+                if (value < 10.0f)
+                {
+                    volumePassThrough = 10.0f;
+                }
+                else if (value > 22000.0f)
+                {
+                    volumePassThrough = 22000.0f;
+                }
+                else
+                {
+                    volumePassThrough = value;
+                }
+            }
+        }
+
+        // Update is not used, but is kept so that this component can be enabled/disabled.
+        private void Update() { }
+
+        /// <summary>
+        /// Applies the audio effect.
+        /// </summary>
+        /// <param name="soundEmittingObject">The GameObject on which the effect is to be applied.</param>
+        public void ApplyEffect(GameObject soundEmittingObject)
+        {
+            if (!isActiveAndEnabled)
+            { return; }
+
+            AudioSource audioSource = soundEmittingObject.GetComponent<AudioSource>();
+            if (audioSource == null)
+            {
+                Debug.LogWarning("The specified emitter does not have an attached AudioSource component.");
+                return;
+            }
+
+            // Audio occlusion is performed using a low pass filter.                
+            AudioLowPassFilter lowPass = soundEmittingObject.GetComponent<AudioLowPassFilter>();
```

Comment:
` var lowPass = soundEmittingObject.EnsureComponent<AudioLowPassFilter>();`

Reply:
As I understand it, this is an optional component, not a required one.

Reply:
If it's optional why do we attempt to get it then add one if it doesn't exist?

Reply:
the low pass filter is not on the occluder, it is on the object containing the AudioInfluencerController. If it does not have this object, we add it

            AudioLowPassFilter lowPass = soundEmittingObject.GetComponent<AudioLowPassFilter>();
            if (lowPass == null)
            { 
                lowPass = soundEmittingObject.AddComponent<AudioLowPassFilter>();
            }

Reply:
Correct, so then we can use the proposed `EnsureComponent`

Reply:
Under the hood the method does exactly what's written:

```
        public static T EnsureComponent<T>(this GameObject gameObject) where T : Component
        {
            T foundComponent = gameObject.GetComponent<T>();
            if (foundComponent == null)
            {
                return gameObject.AddComponent<T>();
            }

            return foundComponent;
        }
```